aten/src/ATen/native/cpu/TensorCompareKernel.cpp

```diff
@@ -213,6 +215,58 @@ static void isneginf_kernel_impl(TensorIterator& iter) {
   });
 }
 
+static void mode_kernel_impl(
+    Tensor& values,
+    Tensor& indices,
+    const Tensor& self,
+    int64_t dim,
+    bool keepdim) {
+  int64_t self_dim_size = ensure_nonempty_size(self, dim);
+
+  AT_DISPATCH_ALL_TYPES_AND(kHalf, self.scalar_type(), "mode_cpu", [&] {
+    compare_base_kernel<scalar_t>(
+        values,
+        indices,
+        self,
+        dim,
+        keepdim,
+        [&](scalar_t* values_data,
+            int64_t* indices_data,
+            const scalar_t* self_data,
+            auto self_dim_stride) {
+          scalar_t mode = 0;
+          int64_t modei = 0;
+          int64_t temp_freq = 0;
+          int64_t max_freq = 0;
+
+          std::vector<std::pair<scalar_t, int64_t>> elements(self_dim_size);
+          std::vector<int64_t> idxs(self_dim_size);
```

Comment:
This will allocate new memory `self.numel() / self.size(dim)` times whereas you only need one to allocate once per thread. I suggest adding a `compare_base_kernel` variant that lets the caller define their own `for_each` loop (line 50).

i.e. 
```cpp
static inline void compare_base_kernel_core(Tensor& result1, Tensor& result2, const Tensor& self, int64_t dim, bool keepdim, const TensorIterator::loop_t& loop);

template <typename scalar_t, typename scalar_t_2=int64_t, typename func_t>
static inline void compare_base_kernel(Tensor& result1, Tensor& result2,
    const Tensor& self,
    int64_t dim,
    bool keepdim,
    const func_t& f) {
  auto loop = ...; // From line 50
  return compare_base_kernel_core(result1, result2, self, dim, keepdim, loop);
}
```

Then you can call `compare_base_kernel_core` from this kernel and allocate the memory only once inside the loop.

Reply:
That makes sense.
Thanks for the suggestion.